noise_corrector.py

```diff
@@ -1,70 +1,177 @@
+  x = randomized_response_rate
 
-  total_records = sum(buckets.values())
-  num_conversion_buckets = len(buckets)
+  # An observed value v will be distributed according to
+  # BinomialDistribution[n, q] where
+  # q = (1 - x)*(v' / n) + xr
+  # Where v' is the true value.
+  #
+  # E[v_hat] = E[(v - (1 - b)nxr) / (1 - (1 - b)x)]
+  # = (v'(1 - x) + nxr - (1 - b)nxr) / (1 - (1 - b)x)
+  # = (v'(1 - x) + bnxr) / (1 - (1 - b)x)
+  #
+  # For b = 0 --> E[v_hat] = v' i.e. the estimator is unbiased.
+  # For b = 1 --> E[v_hat] = v'(1-x) + bnxr
+  #
+  # V[v_hat] = V[(v - (1 - b)nxr) / (1 - (1 - b)x)]
+  # = V[v] / (1 - (1 - b)x)^2
+  # = nq(1-q) / (1 - (1 - b)x)^2
+  #
+  # For b = 0 --> V[v_hat] = nq(1-q) / (1 - x)^2
+  # For b = 1 --> V[v_hat] = nq(1-q)
+  def estimate(v):
+    return (v - (1 - beta) * n * x * r) / (1 - (1 - beta) * x)
 
-  # |noise_probability| of the reports are noised and uniformly distributed
-  # among the conversion buckets so one can calculate how many values have
-  # were from noised, per bucket.
-  noised_values_per_bucket = total_records * noise_probability / num_conversion_buckets
+  return [estimate(v) for v in observed_values]
+
+
+def generate_possible_outputs_per_source(max_reports, data_cardinality,
+                                         max_windows):
+  """Enumerates all possible outputs for any given source event
+
+  max_reports: The maximum number of triggers that can be attributed to this
+                source.
+  data_cardinality: The number of possible trigger_data values for the source.
+  max_windows: The number of reporting windows any given report can show up in.
+
+  Returns: a JSON list of all possible outputs. Each output is a list of size
+           `max_reports`, describing each report (its trigger_data and window).
+  """
+
+  def find_k_comb(N, k):
+    """Computes the Nth lexicographically smallest k-combination
+
+    https://en.wikipedia.org/wiki/Combinatorial_number_system
+
+    Follows the greedy algorithm outlined here:
+    http://math0.wvstateu.edu/~baker/cs405/code/Combinadics.html
+    """
+    target = N
+    c = 0
+    while math.comb(c, k) <= N:
+      c += 1
+
+    combos = []
+    while c >= 0 and k >= 0:
+      coef = math.comb(c, k)
+      if coef <= target:
+        combos.append(c)
+        k -= 1
+        target -= coef
+      c -= 1
+    return combos
+
+  def bars_to_report(star_index, num_star):
+    """Returns a report description"""
+
+    bars_prior_to_star = star_index - (max_reports - 1 - num_star)
+
+    # Any stars before any bars means the report is suppressed.
+    if bars_prior_to_star == 0:
+      return "No report"
+
+    # Otherwise, the number of bars just indexes into the cartesian product
+    # of |data_cardinality| x |num_windows|.
+    # Subtract the first bar because that just begins the first window.
+    window, data = divmod(bars_prior_to_star - 1, data_cardinality)
+    return {"window": window, "data": data}
+
+  def generate_output(N):
```

Comment:
Optional: for the purpose of generating all reports, it might be more efficient to directly enumerate the reports directly without going to stars and bars; that way, star and bar-to-report conversion doesn't have to be applied in its entirely for each possible output.

Reply:
Leaving this as a TODO for now. Efficiency is really not needed here right now, but possibly this could be more readable as a direct enumeration.